:bomb: `Bug`: plugin `jenkins-pipeline` panic when Create - invalid memory address or nil pointer dereference

### What Happened?

- config

```yaml
- name: jenkins-pipeline
  instanceID: default
  dependsOn: []
  options:
    jenkinsURL: http://127.0.0.1:50125
    jenkinsUser: admin
    jenkinsfilePath: https://raw.githubusercontent.com/steinliber/ci-repo/main/springboot/Jenkinsfile
    jenkinsEnableRestart: true
    projectURL: hhttp://127.0.0.1/root/spring-demo
```

- logs

```sh
2022-09-02 10:36:26 ℹ [INFO]  Processing: (jenkins-pipeline/default) -> Create ...
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x2 addr=0x0 pc=0x1434bfa80]

goroutine 1 [running]:
github.com/devstream-io/devstream/pkg/util/gitlab.(*Client).FileExists(0x14000b30050, {0x13594fcf3, 0xb})
        github.com/devstream-io/devstream/pkg/util/gitlab/file.go:42 +0x1c0
github.com/devstream-io/devstream/pkg/util/gitlab.(*Client).PushLocalFileToRepo(0x14000b30050, 0x140010106f0, 0x1)
        github.com/devstream-io/devstream/pkg/util/gitlab/commit.go:20 +0x328
github.com/devstream-io/devstream/internal/pkg/plugininstaller/ci.PushCIFiles(0x14000c7fef0)
        github.com/devstream-io/devstream/internal/pkg/plugininstaller/ci/installer.go:26 +0x22c
github.com/devstream-io/devstream/internal/pkg/plugininstaller.(*Operator).Execute(0x14001090f48, 0x14000c7fef0)
        github.com/devstream-io/devstream/internal/pkg/plugininstaller/plugininstaller.go:68 +0x238
github.com/devstream-io/devstream/internal/pkg/plugin/jenkinspipeline.Create(0x140012070b0)
        github.com/devstream-io/devstream/internal/pkg/plugin/jenkinspipeline/create.go:26 +0x20c
github.com/devstream-io/devstream/cmd/plugin/jenkins-pipeline.Plugin.Create({0x0, 0x0}, 0x140012070b0)
        github.com/devstream-io/devstream/cmd/plugin/jenkins-pipeline/main.go:16 +0x40
github.com/devstream-io/devstream/internal/pkg/pluginengine.Create(0x14000b586c0)
        github.com/devstream-io/devstream/internal/pkg/pluginengine/plugin.go:26 +0xe4
github.com/devstream-io/devstream/internal/pkg/pluginengine.execute({0x102aa89d8, 0x1400000e6d8}, {0x140001a5308, 0x1, 0x1}, 0x0)
        github.com/devstream-io/devstream/internal/pkg/pluginengine/change.go:136 +0x84c
github.com/devstream-io/devstream/internal/pkg/pluginengine.Apply({0x16fcd3827, 0x26}, 0x0)
        github.com/devstream-io/devstream/internal/pkg/pluginengine/cmd_apply.go:55 +0x5d0
main.applyCMDFunc(0x103ae0d00, {0x14000477420, 0x0, 0x2})
        github.com/devstream-io/devstream/cmd/devstream/apply.go:24 +0x94
github.com/spf13/cobra.(*Command).execute(0x103ae0d00, {0x14000477400, 0x2, 0x2})
        github.com/spf13/cobra@v1.4.0/command.go:860 +0x854
github.com/spf13/cobra.(*Command).ExecuteC(0x103ae0f80)
        github.com/spf13/cobra@v1.4.0/command.go:974 +0x6f0
github.com/spf13/cobra.(*Command).Execute(0x103ae0f80)
        github.com/spf13/cobra@v1.4.0/command.go:902 +0x30
main.main()
        github.com/devstream-io/devstream/cmd/devstream/main.go:101 +0x2c
```

### How to Reproduce?

_No response_

### Anything else

_No response_

### DevStream Version

latest

--- a/internal/pkg/plugininstaller/common/repo.go
+++ b/internal/pkg/plugininstaller/common/repo.go
@@ -3,7 +3,6 @@ package common
 import (
 	"fmt"
 	"net/url"
-	"os"
 	"strings"
 
 	"github.com/devstream-io/devstream/pkg/util/file"
@@ -77,17 +76,6 @@ func (d *Repo) BuildRepoInfo() *git.RepoInfo {
 		BaseURL:    d.BaseURL,
 		Type:       d.RepoType,
 	}
-}
-
-// GetRepoToken get repo token from env
-func (d *Repo) GetRepoToken() string {
-	switch d.RepoType {
-	case "github":
-		os.Getenv(github.TokenEnvKey)
-	case "gitlab":
-		os.Getenv(gitlab.TokenEnvKey)
-	}
-	return ""
 }
 
 // BuildURL return url build from repo struct

--- a/internal/pkg/plugininstaller/jenkins/validate.go
+++ b/internal/pkg/plugininstaller/jenkins/validate.go
@@ -56,9 +56,6 @@ func ValidateJobConfig(options plugininstaller.RawOptions) (plugininstaller.RawO
 	if err = validator.StructAllError(opts); err != nil {
 		return nil, err
 	}
-	if token := opts.ProjectRepo.GetRepoToken(); token == "" {
-		return nil, errors.New("git repo token is required")
-	}
 	if opts.ProjectRepo.RepoType == "github" {
 		return nil, errors.New("jenkins job not support github for now")
 	}

--- a/internal/pkg/plugininstaller/plugininstaller.go
+++ b/internal/pkg/plugininstaller/plugininstaller.go
@@ -76,6 +76,9 @@ func (o *Operator) Execute(options RawOptions) (map[string]interface{}, error) {
 	if o.GetStateOperation != nil {
 		log.Debugf("Start to execute GetStateOperation...")
 		state, err = o.GetStateOperation(options)
+		if err != nil {
+			return nil, err
+		}
 	}
 	return state, err
 }

--- a/pkg/util/pkgerror/pluginerror.go
+++ b/pkg/util/pkgerror/pluginerror.go
@@ -1,0 +1,43 @@
+package pkgerror
+
+import (
+	"fmt"
+	"strings"
+)
+
+type ErrorMessage string
+
+// PluginError is plugin's error wrapper
+type PluginError struct {
+	// which plugin the error raises
+	PluginName string
+	// error message
+	Message string
+	// error extra module info
+	Module string
+}
+
+func (e *PluginError) Error() string {
+	moduleMsg := ""
+	if e.Module != "" {
+		moduleMsg = fmt.Sprintf("(%s)", e.Module)
+	}
+	return fmt.Sprintf("[%s]%s: %s", e.PluginName, moduleMsg, e.Message)
+}
+
+func CheckSlientErrorByMessage(targetError error, errCheckStrings ...ErrorMessage) bool {
+	for _, checkErr := range errCheckStrings {
+		if strings.Contains(targetError.Error(), string(checkErr)) {
+			return true
+		}
+	}
+	return false
+}
+
+func NewErrorFromPlugin(pluginName, moduleName string, err error) *PluginError {
+	return &PluginError{
+		PluginName: pluginName,
+		Message:    err.Error(),
+		Module:     moduleName,
+	}
+}

--- a/pkg/util/scm/gitlab/commit.go
+++ b/pkg/util/scm/gitlab/commit.go
@@ -1,10 +1,9 @@
 package gitlab
 
 import (
-	"strings"
-
 	"github.com/xanzy/go-gitlab"
 
+	"github.com/devstream-io/devstream/pkg/util/pkgerror"
 	"github.com/devstream-io/devstream/pkg/util/scm/git"
 )
 
@@ -35,11 +34,8 @@ func (c *Client) PushLocalFileToRepo(commitInfo *git.CommitInfo, checkUpdate boo
 	}
 	createCommitOptions := c.CreateCommitInfo(gitlab.FileCreate, commitInfo)
 	_, _, err := c.Commits.CreateCommit(c.GetRepoPath(), createCommitOptions)
-	if err != nil {
-		if strings.Contains(err.Error(), "A file with this name already exists") {
-			return false, nil
-		}
-		return true, err
+	if err != nil && pkgerror.CheckSlientErrorByMessage(err, errFileExist) {
+		return true, c.newModuleError(err)
 	}
 	return false, nil
 }

--- a/pkg/util/scm/gitlab/errors.go
+++ b/pkg/util/scm/gitlab/errors.go
@@ -1,0 +1,30 @@
+package gitlab
+
+import (
+	"errors"
+	"strings"
+
+	"github.com/devstream-io/devstream/pkg/util/pkgerror"
+)
+
+var (
+	errRepoNotFound   pkgerror.ErrorMessage = "Project Not Found"
+	errRepoExist      pkgerror.ErrorMessage = "name already exists"
+	errWebHookInvalid pkgerror.ErrorMessage = "invlid url given"
+	errFileExist      pkgerror.ErrorMessage = "A file with this name already exists"
+)
+
+var errorMsgMap = map[pkgerror.ErrorMessage]string{
+	errWebHookInvalid: "webhook config doesn't support local network, should config gitlab or change jenkinsURL config",
+	errFileExist:      "file already exist",
+}
+
+func (c *Client) newModuleError(err error) error {
+	var newError = err
+	for k, v := range errorMsgMap {
+		if strings.Contains(err.Error(), string(k)) {
+			newError = errors.New(v)
+		}
+	}
+	return pkgerror.NewErrorFromPlugin("gitlab", c.GetRepoPath(), newError)
+}

--- a/pkg/util/scm/gitlab/file.go
+++ b/pkg/util/scm/gitlab/file.go
@@ -2,11 +2,11 @@ package gitlab
 
 import (
 	"net/http"
-	"strings"
 
 	"github.com/xanzy/go-gitlab"
 
-	"github.com/devstream-io/devstream/pkg/util/log"
+	mapset "github.com/deckarep/golang-set/v2"
+
 	"github.com/devstream-io/devstream/pkg/util/scm/git"
 )
 
@@ -14,11 +14,7 @@ func (c *Client) DeleteFiles(commitInfo *git.CommitInfo) error {
 	deleteCommitoptions := c.CreateCommitInfo(gitlab.FileDelete, commitInfo)
 	_, _, err := c.Commits.CreateCommit(c.GetRepoPath(), deleteCommitoptions)
 	if err != nil {
-		if strings.Contains(err.Error(), "404") {
-			log.Debugf("gitlab repo has been deleted")
-			return nil
-		}
-		return err
+		return c.newModuleError(err)
 	}
 	return nil
 }
@@ -27,7 +23,7 @@ func (c *Client) UpdateFiles(commitInfo *git.CommitInfo) error {
 	updateCommitoptions := c.CreateCommitInfo(gitlab.FileUpdate, commitInfo)
 	_, _, err := c.Commits.CreateCommit(c.GetRepoPath(), updateCommitoptions)
 	if err != nil {
-		return err
+		return c.newModuleError(err)
 	}
 	return nil
 }
@@ -58,11 +54,10 @@ func (c *Client) GetLocationInfo(path string) ([]*git.RepoFileStatus, error) {
 	}
 
 	file, response, err := c.RepositoryFiles.GetFile(c.GetRepoPath(), path, getFileOptions)
+	errCodeSet := mapset.NewSet(http.StatusBadRequest, http.StatusUnauthorized, http.StatusNotFound)
 
-	for _, v := range []int{http.StatusBadRequest, http.StatusUnauthorized, http.StatusNotFound} {
-		if response.StatusCode == v {
-			return gitRepoFileStatus, nil
-		}
+	if response != nil && errCodeSet.Contains(response.StatusCode) {
+		return gitRepoFileStatus, nil
 	}
 
 	if err != nil {

--- a/pkg/util/scm/gitlab/repo.go
+++ b/pkg/util/scm/gitlab/repo.go
@@ -1,11 +1,10 @@
 package gitlab
 
 import (
-	"strings"
-
 	"github.com/xanzy/go-gitlab"
 
 	"github.com/devstream-io/devstream/pkg/util/log"
+	"github.com/devstream-io/devstream/pkg/util/pkgerror"
 	"github.com/devstream-io/devstream/pkg/util/scm/git"
 )
 
@@ -51,11 +50,7 @@ func (c *Client) InitRepo() error {
 		p.NamespaceID = gitlab.Int(groupId)
 	}
 	_, _, err = c.Projects.CreateProject(p)
-	if err != nil {
-		if strings.Contains(err.Error(), "has already been taken") {
-			log.Debugf("gitlab repo %s already exist, not create...", c.Repo)
-			return nil
-		}
+	if err != nil && pkgerror.CheckSlientErrorByMessage(err, errRepoNotFound) {
 		return err
 	}
 
@@ -64,10 +59,9 @@ func (c *Client) InitRepo() error {
 
 func (c *Client) DeleteRepo() error {
 	_, err := c.Projects.DeleteProject(c.GetRepoPath())
-	if err != nil {
+	if err != nil && pkgerror.CheckSlientErrorByMessage(err, errRepoNotFound) {
 		return err
 	}
-
 	return nil
 }
 
@@ -76,8 +70,8 @@ func (c *Client) DescribeRepo() (*gitlab.Project, error) {
 	res, _, err := c.Projects.GetProject(c.GetRepoPath(), p)
 
 	if err != nil {
-		log.Debugf("gitlab project: get [%s] info error %s", c.GetRepoPath(), err)
-		return nil, err
+		log.Debugf("gitlab project: [%s] error %+v", c.GetRepoPath(), err)
+		return nil, c.newModuleError(err)
 	}
 
 	return res, nil
@@ -88,30 +82,35 @@ func (c *Client) AddWebhook(webhookConfig *git.WebhookConfig) error {
 	if err != nil {
 		return err
 	}
-	if projectHook == nil {
-		p := &gitlab.AddProjectHookOptions{
-			PushEvents: gitlab.Bool(true),
-			Token:      gitlab.String(webhookConfig.SecretToken),
-			URL:        gitlab.String(webhookConfig.Address),
-		}
-		_, _, err := c.Projects.AddProjectHook(c.GetRepoPath(), p)
-		return err
+	if projectHook != nil {
+		log.Debugf("gitlab AddWebhook already exist")
+		return nil
 	}
-	log.Debugf("gitlab AddWebhook already exist")
+	p := &gitlab.AddProjectHookOptions{
+		PushEvents: gitlab.Bool(true),
+		Token:      gitlab.String(webhookConfig.SecretToken),
+		URL:        gitlab.String(webhookConfig.Address),
+	}
+	_, _, err = c.Projects.AddProjectHook(c.GetRepoPath(), p)
+	if err != nil {
+		return c.newModuleError(err)
+	}
 	return nil
 }
 
 func (c *Client) DeleteWebhook(webhookConfig *git.WebhookConfig) error {
 	projectHook, err := c.getWebhook(webhookConfig)
+	if err != nil && pkgerror.CheckSlientErrorByMessage(err, errRepoExist) {
+		return err
+	}
+	if projectHook == nil {
+		log.Debugf("gitlab DeleteWebhook not found")
+		return nil
+	}
+	_, err = c.Projects.DeleteProjectHook(c.GetRepoPath(), projectHook.ID)
 	if err != nil {
-		log.Debugf("gitlab DeleteWebhook list hooks failed: %s", err)
-		return err
+		return c.newModuleError(err)
 	}
-	if projectHook != nil {
-		_, err := c.Projects.DeleteProjectHook(c.GetRepoPath(), projectHook.ID)
-		return err
-	}
-	log.Infof("gitlab DeleteWebhook not found")
 	return nil
 }
 
@@ -119,8 +118,8 @@ func (c *Client) getWebhook(webhookConfig *git.WebhookConfig) (*gitlab.ProjectHo
 	p := &gitlab.ListProjectHooksOptions{}
 	hooks, _, err := c.Projects.ListProjectHooks(c.GetRepoPath(), p)
 	if err != nil {
-		log.Debugf("gitlab DeleteWebhook lsit hooks failed: %s", err)
-		return nil, err
+		log.Debugf("gitlab DeleteWebhook list hooks failed: %s", err)
+		return nil, c.newModuleError(err)
 	}
 	for _, hook := range hooks {
 		if hook.URL == webhookConfig.Address {

